amp-gwd-animation 0.1 (initial)

sklobovskaya opened the pull request:
Initial submission for the amp-gwd-animation extension (https://github.com/ampproject/amphtml/issues/9949).

- The extension in experiment mode
- A service associated with the extension
- Unit tests for the extension and service
- Validator rules

honeybadgerdontcare commented on extensions/amp-gwd-animation/0.1/validator-amp-gwd-animation.protoascii:
```diff
@@ -0,0 +1,39 @@
+#
```
FYI, validator rules are now at the extension level, not version level. Please move up one directory.

sklobovskaya replied:
Done.

alanorozco commented on extensions/amp-gwd-animation/0.1/amp-gwd-animation-impl.js:
```diff
@@ -0,0 +1,447 @@
+    this.playLabelAnimation_(receiver, label);
+
+    // Trigger a pause.
+    const timeToWaitBeforePause = 40;
```
Why 40ms specifically?

Could this be a module-level constant?

sklobovskaya replied:
Yes, this should be a constant. Done.

The 40ms delay is necessary to work around a still-present Safari bug (in our regular non-AMP runtime, too). I added some comments in the code to mention this.

dvoytenko commented on extensions/amp-gwd-animation/0.1/amp-gwd-animation-impl.js:
```diff
@@ -0,0 +1,447 @@
+    }
+
+    if (!eventName) {
+      user().warn(LOG_ID, `Invalid event name ${eventName}.`);
```
Why warning? Should we just do an error?

sklobovskaya replied:
Yes, these can just be errors. Done.

dvoytenko commented on extensions/amp-gwd-animation/0.1/amp-gwd-animation-impl.js:
```diff
@@ -0,0 +1,447 @@
+        `${GWD_TIMELINE_EVENT}`,
+        {eventName: userEventName, sourceEvent: event});
+
+    this.ampdoc_.win.dispatchEvent(timelineEvent);
```
This could be tricky. Where will the event go in this case? Will we avoid it being shared to other embeds or documents?

sklobovskaya replied:
Right, this event should just be scoped to the ampdoc root. Done.

dvoytenko commented on extensions/amp-gwd-animation/0.1/amp-gwd-animation-impl.js:
```diff
@@ -0,0 +1,447 @@
+ * GWD_SERVICE_NAME.
+ * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+ */
+export const installGwdRuntimeServiceForDoc = function(ampdoc) {
```
This is optional for now, but may be needed if you plan for this service to be used w/o the `amp-gwd-animation` element.

This changed a bit, and now looks like this: https://github.com/ampproject/amphtml/blob/db824d0f7f42d75444df1e97461d789819481faf/extensions/amp-dynamic-css-classes/0.1/amp-dynamic-css-classes.js#L155

Could you please change? That usually means that you may also need to include extension explicitly in test like this: https://github.com/ampproject/amphtml/blob/master/extensions/amp-3q-player/0.1/test/test-amp-3q-player.js#L25

On the plus side this provides for better test isolation.

sklobovskaya replied:
Updated to use registerServiceForDoc().

dvoytenko commented on extensions/amp-gwd-animation/0.1/amp-gwd-animation.css:
```diff
@@ -0,0 +1,22 @@
+/**
+ * Class used to disable GWD animations.
+ */
+.i-amphtml-gwd-animation-disabled [class*=gwdanimation] {
```
Just to confirm: this is your normal practice as well, right?

sklobovskaya replied:
Which part are you referring to? Our regular non-AMP runtime does not have need to disable animations globally, so we don't employ any CSS quite like this.

dvoytenko replied:
I see. Did you try `.i-amphtml-gwd-animation-disabled * {animation:none}` and found it less performant?

sklobovskaya replied:
Leaving as-is for now as discussed.

dvoytenko commented on extensions/amp-gwd-animation/0.1/amp-gwd-animation.js:
```diff
@@ -0,0 +1,228 @@
+    // Walk the invocation object to get the requested property by
+    // its path, e.g., 'args.id' == invocation.args.id
+    let obj = invocation;
+    const props = argDef.split('.');
```
Please see if this utility does the same for you: https://github.com/ampproject/amphtml/blob/master/src/json.js#L82

sklobovskaya replied:
Yes, that works great. Thanks!

dvoytenko commented on extensions/amp-gwd-animation/0.1/amp-gwd-animation.js:
```diff
@@ -0,0 +1,228 @@
+
+    // Listen for GWD timeline events to re-broadcast them via the doc action
+    // service.
+    this.getAmpDoc().win.addEventListener(
```
Same question here. Is it possible to catch events on the root node instead?

sklobovskaya replied:
Done.

dvoytenko commented on extensions/amp-gwd-animation/0.1/amp-gwd-animation.js:
```diff
@@ -0,0 +1,228 @@
+ * @param {string} actionStr e.g., `someDiv.hide`
+ * @private Visible for testing.
+ */
+export const insertEventActionBinding = function(element, event, actionStr) {
```
Ditto on function forms

sklobovskaya replied:
Done.

dvoytenko commented on extensions/amp-gwd-animation/0.1/amp-gwd-animation.js:
```diff
@@ -0,0 +1,228 @@
+        actionStr;
+  }
+
+  element.setAttribute('on', newOnAttrVal);
```
Just fyi, you might be also able to use `actionService.installActionHandler(...)`

sklobovskaya replied:
Does that entirely replace the handler for all actions on the element? I think I need the default AMP action handler to remain, since it may have other event/action definitions that I don't want to have to handle myself, but simply add an additional action def to the list. I couldn't find another way to do it except to manually manipulate the `on` attribute value. Let me know if there's a better way.

dvoytenko replied:
This is a bit racy because actions could be cached. Can we maybe change this a bit via calling a method on `actionService` that will do `element.setAttribute('on', ...)` but will also reset the cached value?

sklobovskaya replied:
Added addAction method as discussed.

sklobovskaya commented on yarn.lock:
```diff
@@ -7569,6 +7569,10 @@ sax@^1.1.4, sax@~1.2.1:
   version "1.2.1"
   resolved "https://registry.yarnpkg.com/sax/-/sax-1.2.1.tgz#7b8e656190b228e81a66aea748480d828cd2d37a"
 
+seedrandom@^2.4.2:
```
Done.

sklobovskaya commented on validator/nodejs/package.json:
```diff
@@ -2,7 +2,13 @@
   "version": "1.0.20",
   "description": "Official validator for AMP HTML (www.ampproject.org)",
-  "keywords": ["AMP", "validator", "validate", "AMP HTML", "Accelerated Mobile Pages"],
+  "keywords": [
```
Done.

sklobovskaya commented on tools/experiments/experiments.js:
```diff
@@ -106,6 +106,12 @@ const EXPERIMENTS = [
     cleanupIssue: 'https://github.com/ampproject/amphtml/issues/3996',
   },
   {
+    id: 'amp-gwd-animation',
```
Done.

sklobovskaya commented on extensions/amp-gwd-animation/0.1/amp-gwd-animation-impl.js:
```diff
@@ -0,0 +1,447 @@
+ * @param {string} counterName
+ * @private
+ */
+const getOrInitCounter = function(receiver, counterName) {
```
Done throughout.

sklobovskaya commented on extensions/amp-gwd-animation/0.1/amp-gwd-animation-impl.js:
```diff
@@ -0,0 +1,447 @@
+  setEnabled(enable) {
+    if (enable) {
+      // Lift the animation CSS override.
+      this.ampdoc_.getBody().classList.remove(ANIMATIONS_DISABLED_CLASS);
```
Done.

sklobovskaya commented:
@dvoytenko @alanorozco Addressed all comments up to this point. Please take another look. Thanks!

Note: Some visual diffs cropped up on this last push; I wasn't expecting any of my changes to cause it, but let me know if something inadvertently did so. I'm not authorized to see output from the percy checks myself.

dvoytenko commented on extensions/amp-gwd-animation/0.1/amp-gwd-animation-impl.js:
```diff
@@ -0,0 +1,426 @@
+
+    const timelineEvent = createCustomEvent(
+        this.ampdoc_.win,
+        `${GWD_TIMELINE_EVENT}`,
```
Just ``${GWD_TIMELINE_EVENT}` == GWD_TIMELINE_EVENT`?

sklobovskaya replied:
Done.

dvoytenko commented:
LGTM on my side.

@alanorozco Please complete review as well.

alanorozco commented on extensions/amp-gwd-animation/0.1/amp-gwd-animation-impl.js:
```diff
@@ -0,0 +1,426 @@
+ * @private
+ */
+function getCounter(receiver, counterName) {
+  if (receiver.gwdGotoCounters &&
```
Nit: you could short-circuit:

```javascript
if (receiver.gwdGotoCounters &&
    receiver.gwdGotoCounters.hasOwnProperty(counterName)) {
  return receiver.gwdGotoCounters[counterName];
}
return 0;
```

sklobovskaya replied:
Done.

alanorozco commented on extensions/amp-gwd-animation/0.1/amp-gwd-animation-impl.js:
```diff
@@ -0,0 +1,447 @@
+   * @private
+   */
+  initialize_() {
+    // TODO: The GWD animation runtime should start out disabled, but leaving it
```
@sklobovskaya Could you please add `(#7846)` to the TODO comment?

sklobovskaya commented:
@alanorozco Done with the last round of changes. Thanks!

sklobovskaya commented:
@alanorozco @dvoytenko Anything still needed on my end on this PR, or can it be merged?

dvoytenko commented:
I think all LGTMs are in. Just trying to get Travis CI pass.

dvoytenko commented:
Merged! Thanks a lot @sklobovskaya !
